pybaseball/analysis/trajectories/batted_balls/calculator.py

```diff
@@ -0,0 +1,245 @@
+        launch_angle,
+        launch_direction_angle,
+        initial_spin,
+        spin_angle,
+        delta_time=0.01,
+    ):
+        # TODO: make the return value a trajectory object
+        """
+        computes a batted ball trajectory. speed is in miles-per-hour,
+        angles in degrees, and spin in revolutions per minute
+
+        :param initial_speed: float
+        :param launch_angle: float
+        :param launch_direction_angle: float
+        :param initial_spin: float
+        :param spin_angle: float
+        :param delta_time: float
+        :return: pandas data frame
+        """
+
+        initial_velocity = (
+            initial_speed
+            * self.env_parameters.unit_conversions.mph_to_fts
+            * np.array(
+                [
+                    cos_in_degrees(launch_angle)
+                    * sin_in_degrees(launch_direction_angle),
+                    cos_in_degrees(launch_angle)
+                    * cos_in_degrees(launch_direction_angle),
+                    sin_in_degrees(launch_angle),
+                ]
+            )
+        )
+
+        initial_conditions = np.concatenate(
+            (self.initial_position, initial_velocity), axis=0
+        )
+
+        rk_solution = RK45(
+            partial(
+                self.trajectory_fun,
+                launch_angle=launch_angle,
+                launch_direction_angle=launch_direction_angle,
+                spin=initial_spin,
+                spin_angle=spin_angle,
+            ),
+            0,
+            initial_conditions,
+            t_bound=1000,
+            max_step=delta_time,
+        )
+        ans = []
+        z = self.initial_position[2]
+        while z >= 0:
+            rk_solution.step()
+            res = rk_solution.y
+            z = res[2]
+            ans.append([rk_solution.t] + list(res))
+        result_df = pd.DataFrame(np.array(ans).reshape(-1, 7))
+        result_df.columns = ["t", "x", "y", "z", "vx", "vy", "vz"]
+        return result_df
+
+    def trajectory_fun(
+        self,
+        t,
+        trajectory_vars,
+        spin=2500,
+        spin_angle=0,
+        launch_angle=0,
+        launch_direction_angle=0,
+    ):
+        """
+        function for computing the trajectory using the 4th-order Runge-Kutta method.
+        trajectory vars are the 3 positions and 3 velocity components of the ball.
+        returns the derivatives of the input variables, i.e., the 3 velocity components,
+        and the 3 acceleration components.
+
+        :param t: float
+        :param trajectory_vars: tuple(float)
+        :param spin: float
+        :param spin_angle: float
+        :param launch_angle: float
+        :param launch_direction_angle: float
+        :return: numpy array
+        """
+        # trajectory_vars = x, y, z, vx, vy, vz
+        _, _, _, vx, vy, vz = trajectory_vars
+        v = np.sqrt(vx ** 2 + vy ** 2 + vz ** 2)
+
+        sidespin = spin * sin_in_degrees(spin_angle)
```

Comment:
Would it make sense to just convert spin_angle and launch_angle to radians once and then streamline the rest of these calls?

Reply:
What I did here was add `lru_cache` decorators to `sin_in_degrees` and `cos_in_degrees`. And internally they call a cached conversion from degrees to radians. This way the conversion only happens once, which I think captures the optimization you're referring to in this comment?

Reply:
I'm less bothered about the runtime cost. It's more that it'd be a lot cleaner I think to shuttle everything into radians once and then just have `sin`, `cos`, etc.

Reply:
if you look at where the `sin_in_degrees` and `cos_in_degrees` were being used it was basically two places - making a velocity vector from the launch angles, and decomposing the spin into x, y, z components. To address this comment, I refactored these both into `utils`. so now the trajectpry calculator doesn't mess with angles at all, it just calls `unit_vector` and `spin_components`, and inside of those I took your advice and converted to radians once, and then called the numpy `sin` and `cos`.